File: FSharp/5_Natural_Transformation.fs
module Natural_Transformation

open FsCheck.Xunit
open Xunit

let maybeHead = function
    | [] -> None
    | head::_ -> Some head
    
[<Fact>]
let transform_empty_list =
    Assert.Equal(maybeHead [], None)
    
[<Property>]
let transform_non_empty_list (head : int) tail =
    maybeHead (head::tail) = Some head

File: CSharp/5_Natural_Transformation.cs
using System.Collections.Generic;
using System.Linq;
using FsCheck;
using Xunit;
// ReSharper disable All

namespace CSharp
{
    public class Natural_Transformation
    {
        public static Maybe<T> MaybeHead<T>(List<T> list) =>
            list.Any() ? Maybe<T>.Some(list.First()) : Maybe<T>.None();

        [Fact]
        public void Transform_empty_list()
        {
            Assert.Equal(MaybeHead(new List<int>()), Maybe<int>.None());
        }

        [Fact]
        public void Transform_non_empty_list()
        {
            Prop.ForAll(
                    Arb.From<int>(),
                    Arb.From<List<int>>(),
                    (head, tail) => 
                        MaybeHead(tail.Prepend(head).ToList()).Equals(Maybe<int>.Some(head)))
                .QuickCheckThrowOnFailure();
        }
    }
}